histogram.c

```diff
@@ -0,0 +1,92 @@
+//---------------------------------------------------------------------------------------------
+//
+// Copyright (c) 2018, fmad engineering llc 
+//
+// The Creative Commons BY-NC 4.0 International License see LICENSE file for details 
+// 
+// Histogram generation from pcap
+//
+//---------------------------------------------------------------------------------------------
+
+#include <stdio.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <string.h>
+#include <malloc.h>
+#include <errno.h>
+#include <sys/stat.h>
+#include <sys/mman.h>
+#include <sys/shm.h>
+#include <fcntl.h>
+#include <locale.h>
+#include <linux/sched.h>
+#include <pthread.h>
+
+#include "fTypes.h"
+#include "fProfile.h"
+#include "histogram.h"
+
+
+PacketInfoBulk_t* PktInfoBulk_Alloc(u32 MaxPkts)
+{
+	PacketInfoBulk_t *p = malloc(sizeof(PacketInfoBulk_t));
+
+	assert(p != NULL);
+
+	p->Max		= MaxPkts;
+	p->Head		= 0;
+	p->PktInfo	= malloc(sizeof(PacketInfo_t) * p->Max);
+	p->Next		= NULL;
+
+	return p;
+}
+
+void PktInfo_Insert(PacketInfoBulk_t *P, uint16_t Len, uint32_t Tdiff)
+{
+	assert(P != NULL);
+
+	PacketInfoBulk_t *last = P;
+
+	for (; last->Next != NULL; last = last->Next);
```

Comment:
this is going to be extremely slow when inserting 1M packets into a flow.  As PacketInfo_t is a fixed size, how about adding to PacketInfoBulk_t a Pos field

e.g.
PacketInfoBulk_t->PktInfo[ PacketInfoBulk_t->Pos++] when inserting / reading out. Same goes for deciding if its full. 

Pos == Max  

Slightly more complicated but alot faster?

Reply:
I have used "Head" which is the same as "Pos" you are talking about. But there is linked-list on top of the list.
I will make this the other way, so we don't need to travel linked-list while insert.